config/ImathConfig.h.in

```diff
@@ -174,4 +174,7 @@
 #    endif
 #endif
 
+// Whether the copy and move constructors of Vec2, Vec3 and Vec4 are default or not.
+#cmakedefine IMATH_VEC_USE_DEFAULT_CONSTRUCTOR
```

Comment:
I think instead we want

```suggestion
#ifndef IMATH_VEC_USE_DEFAULT_CONSTRUCTOR
#define IMATH_VEC_USE_DEFAULT_CONSTRUCTOR 1
#endif
```

Because it's easy to imagine that a particular downstream app needs to make sure it's the other way, and this would allow it to be overridden if the downstream modules defines the symbol as 0 before including any imath headers.

Then the various places where you have 

    #ifdef IMATH_VEC_USE_DEFAULT_CONSTRUCTOR

should be

    #if IMATH_VEC_USE_DEFAULT_CONSTRUCTOR

Reply:
fly in the ointment, if we go both ways, are we inviting an recurring support burden along these lines?



I can't make my build work with MegaDCC", 

oh right, did you match IMATH_VEC_USE_DEFAULT_CONSTRUCTOR to what they picked? 

but that's in conflict to our internal build

yeah but did you namespace?

we don't want to have to have two copies

Reply:
The OPENEXR_DLL thing was sort of a harbinger, I'm worried about a matrix of macros people have to solve for compatibility

or maybe I'm misunderstanding completely, and it's all inlined and completely doesn't matter which you pick

Reply:
I think because it's inlined it's all ok? There shouldn't be any link issues.

Reply:
If that's the case, great. Then my only request is that we document why we have  it both ways, even if it is a copy and paste of @AlexMWells thorough notes above. This is strongly in the realm of "gosh, this is rather esoteric, why is this a choice, and how do I make it?" I'd prefer the doc is inline in the CMakeLists.txt where you are faced with making a choice.

Reply:
to put it another way, I'm not completely sold that someone will want to exercise this option in practice, and maybe we'd better not if'defing it. I'm just having a hard time understanding why wouldn't need unit tests covering both, and so on. If we're going to treat it thoroughly, is it really worth doing it both ways?

Reply:
Nobody needs an imath-build-time option for this. I argue that it's helpful for downstream-build-time to hide those definitions. It would be extremely hard to make an imath-side unit test for it, but OSL needs to have a way to be sure that we haven't changed imath types so that they won't wreck simd vectorization when used in loops.

Reply:
An example, well with an older compiler, but just wanted to show the behavior.
https://godbolt.org/z/rEY4Ef5GM
With `#define VEC_ENABLE_DEFAULT_CONSTRUCTOR 1`
The optimization report shows:
  ```
   remark #15518: Structure assignment was serialized   [ <source>(145,14) ]
   remark #15518: Structure assignment was serialized   [ <source>(147,14) ]
   remark #15518: Structure assignment was serialized   [ <source>(149,14) ]
   remark #15301: OpenMP SIMD LOOP WAS VECTORIZED
   remark #15450: unmasked unaligned unit stride loads: 11 
   remark #15451: unmasked unaligned unit stride stores: 4 
   remark #15475: --- begin vector cost summary ---
   remark #15476: scalar cost: 367 
   remark #15477: vector cost: 125.870 
   remark #15478: estimated potential speedup: 2.490 
   remark #15488: --- end vector cost summary ---
   ...
   <source>(145,14):remark #34000: call to memcpy implemented inline with loads and stores with proven source 
```
Structure assignment was serialized really kills performance, and we can see the compiler by "default" decided to generate a memcpy. 

By providing the copy constructor and assignment operator explicitly we avoid the memcpy or memset (for initializations) and with #define VEC_ENABLE_DEFAULT_CONSTRUCTOR 0
   The optimization report shows:
```
   remark #15309: vectorization support: normalized vectorization overhead 0.352
   remark #15301: OpenMP SIMD LOOP WAS VECTORIZED
   remark #15450: unmasked unaligned unit stride loads: 11 
   remark #15451: unmasked unaligned unit stride stores: 4 
   remark #15475: --- begin vector cost summary ---
   remark #15476: scalar cost: 201 
   remark #15477: vector cost: 44.370 
   remark #15478: estimated potential speedup: 3.350 
```   

Now this is all very dependent on compiler and order of optimizations, and perhaps isn't as required on newer compilers.  But this is just a simple example, real world kernels could have much larger state being kept on the stack and any nested aggregate that happens to use a memcpy or memset by default could prevent the SROA pass from fully succeeding.

Thus it would be nice for projects like OSL to choose to disable the use of default constructors.  And if the supported compiler matrix has moved on to where it is not a problem then it could chose to enable them.  Problem is examining optimization reports before and after is required.

Reply:
So that sounds like explicitly defining copy and assignment, and not using default, we get the behavior that we want, which is that things optimize as we wish. "we" includes OSL. I can't think of a reason to provide this as one option, and default constructors and copy another option.

I am rereading this argument which suggests maybe we need the option, 

https://github.com/AcademySoftwareFoundation/Imath/pull/337#issuecomment-1644269775

but I'm not understanding a case to preserve the default option. Sorry if I'm being dense....

Reply:
The reason why I made this PR is because I want to copy `Vec3` sequences between host and device using `thrust::async::copy`. It requires `Vec3` to be trivially relocatable, so default constructors and assignments for copy and move are needed for me.
I'm sorry for not telling it.

Reply:
My apologies, I didn't mean to shut down your effort. Maybe I am misunderstanding? I thought that adding the explicit copy and move gives us the trivial relocatability you are searching for, and also satisfies the SROA concern that @AlexMWells raises? I thought we can solve both your need, and also the SIMD concern at the same time, without needing to have a special conditional. My intent wasn't to stop your effort, I was hoping we can find one solution that solves both problems.

Reply:
At first, I apologize for my quick but rough decision.
I misunderstood I couldn't solve the assertion error, but I noticed that it was caused by ODR violation. Though I'm running CI in my forked repository, it will be OK.

I think that the trivial relocatability and the SROA are incompatible because the trivial relocatability requires `Vec3` not to have user-provided copy constructor and assignment operator(even though `Vec3::Vec3(const Vec3& v) : x(v.x), y(v.y), z(v.y) {}`), but on the other hand, providing the copy constructor and assignment operator increases the performance as @AlexMWells showed earlier.
So, we have to switch `default` or not in some way.

Thank you :)